File: java/src/main/java/br/com/emendes/problems/SortColors.java
package br.com.emendes.problems;

/**
 * Dado um array com n objetos coloridos (vermelho, azul e branco), ordene eles in-place (operar direto na entrada)
 * para que objetos da mesma cor sejam adjacentes, com as cores na ordem vermelho, branco e azul.<br><br>
 * <p>
 * Usaremos os inteiros 0, 1 e 2 para representar a cor vermelha, branca e azul, respectivamente.<br><br>
 * <p>
 * Você deve resolver este problema sem usar a função de classificação da biblioteca.<br><br>
 * <p>
 * Restrições:<br>
 * n == nums.length<br>
 * 1 <= n <= 300<br>
 * nums[i] é 0, 1 ou 2.
 */
public class SortColors {

  public void sortColors(int[] nums) {
    int n = nums.length;

    for (int i = 0; i < n; i++) {
      int aux = i;
      for (int j = i + 1; j < n; j++) {
        if (nums[j] < nums[aux]) {
          aux = j;
        }
      }
      int value = nums[i];
      nums[i] = nums[aux];
      nums[aux] = value;
    }
  }

}

File: javascript/problems/SortColors.js
/**
 * Dado um array nums com n objetos coloridos com vermelho, branco ou azul, ordene-os in-place de modo
 * que objetos da mesma cor estejam adjacêntes, com as corres na ordem vermelho, branco e azul.
 *
 * Nós iremos usar os inteiros 0, 1 e 2 para representar as cores vermelho, branco e azul, respectivamente.
 *
 * Você deve revolver esse problema sem usar bibliotecas de ordenação.
 *
 * Restrições:
 *
 * n == nums.length
 *
 * 1 <= n <= 300
 *
 * nums[i] é 0, 1 ou 2
 *
 * Follow-up: Você poderia resolver esse problema percorrendo apenas uma vez o array e
 * com espaço extra constante?
 */
const sortColors = function (nums) {
  let left = 0;
  let right = nums.length - 1;
  let index = 0;

  while (index <= right) {
    if (nums[index] === 2) {
      swap(nums, right, index);
      right--;
    } else {
      if (nums[index] === 0) {
        swap(nums, left, index);
        left++;
      }
      index++;
    }
  }
};

const swap = (arr, i, j) => {
  const memo = arr[i];
  arr[i] = arr[j];
  arr[j] = memo;
};
